### PR TITLE
Calculate covariance matrix in parmest

### LINKED ISSUE #1259: Feature Request: Compute covariance matrix with ParmEst
Is it on the roadmap to compute the covariance matrix in ParmEst based on a local linearization? If not, I may contribute this during the next few weeks.

### PR DESCRIPTION
**Pull request by adowling2**
## Summary/Motivation:
- Estimating the covariance matrix is important for parameter estimation

## Changes proposed in this PR:
- Calculate the covariance matrix (inverse of the reduced hessian) using new capabilities added to contrib/interior_point that leverage pynumero
- This required extending the pysp interface that parmest uses to solve the extensive form with ipopt

## Known/outstanding issues
- [x] For the Rooney-Biegler test problem, all of the covariance matrix elements are too small by a factor of 2.13. This is still under investigation
- [x] Someone familiar with the pysp/parmest interface should look at the proposed changes. Is there a better way to do this? @DLWoodruff @kaklise 
- [x] Someone familiar with pynumero should look at the proposed changes. Is there a cleaner way to integrate the reduced hessian calculation with pysp? @michaelbynum @carldlaird 
- [x] Need to add a test for the Rooney-Biegler example
- [x] Add documentation

## Planned enhancements (future PRs)
- Add visualization of the eigenvectors of the reduced Hessian. This is important for identifiability analysis/diagnostics.

### Legal Acknowledgement

By contributing to this software project, I have read the [contribution guide](https://pyomo.readthedocs.io/en/stable/contribution_guide.html) and agree to the following terms and conditions for my contribution:

1. I agree my contributions are submitted under the BSD license.
2. I represent I am authorized to make the contributions and grant the license. If my employer has rights to intellectual property that includes these contributions, I represent that I have received permission to make contributions and grant the required license on behalf of that employer.

### CODE REVIEW

**Comment by adowling2** [comment]
@DLWoodruff This is an excellent idea. I am working on modifications. I plan to refactor ``parmest`` to no longer call ``solve_ef`` in ``rapper.py`` of ``pysp``.

**Comment by adowling2** [comment]
I implemented the charges suggested by @DLWoodruff. This pull request now does not modify ``rapper.py`` in ``pysp``, except for removing some extra white space on one line.

**Comment by adowling2** [comment]
I finished adding a paragraph to the documentation (with warnings about assumptions!!!) and automated tests. This PR is ready for review. I welcome feedback and suggestions.

File: pyomo/contrib/parmest/parmest.py [hunk 1]
```diff
@@ -31,7 +31,9 @@
 import pyomo.contrib.parmest.ipopt_solver_wrapper as ipopt_solver_wrapper
```
Comment (carldlaird) [inline pyomo/contrib/parmest/parmest.py:31]: I don't really know enough about parmest to judge this part - If @DLWoodruff and/or @kaklise are fine with it, then I am fine with it.

File: doc/OnlineDocs/contributed_packages/parmest/driver.rst [hunk 2]
```diff
@@ -144,3 +144,13 @@ include `model` and `data` and the objective function returns a Pyomo
...
+
+Covariance matrix estimation
+----------------------------
+
+If the optional argument ``calc_cov=True`` is specified for ``theta_est()``, parmest will calculate the covariance matrix :math:`V_{\theta}` as follows:
+
+.. math::
+   V_{\theta} = 2 \sigma^2 H^{-1} 
+
+This formula assumes all measurement errors are independent and identically distributed with variance :math:`\sigma^2`. :math:`H^{-1}` is the inverse of the Hessian matrix for a unweighted sum of least squares problem. Currently, the covariance approximation is only valid if the objective given to parmest is the sum of squared error. Moreover, parmest approximates the variance of the measurement errors as :math:`\sigma^2 = \frac{1}{n-l} \sum e_i^2` where :math:`n` is the number of data points, :math:`l` is the number of fitted parameters, and :math:`e_i` is the residual for experiment :math:`i`.
```
[the remark this answers is not in the document]
Comment (adowling2) [reply doc/OnlineDocs/contributed_packages/parmest/driver.rst]: Yes, right now we are treating sigma as a scalar. IMO, the parmest abstraction needs to be extended to consider more regression cases. #1477

File: pyomo/contrib/parmest/examples/rooney_biegler/parmest_example_cov.py [hunk 3]
```diff
@@ -0,0 +1,39 @@
...
+    return expr
+
+
+solver_options = {"max_iter": 6000}  # not really needed in this case
+
+pest = parmest.Estimator(rooney_biegler_model, data, theta_names, SSE, solver_options)
+obj, theta, cov = pest.theta_est(calc_cov=True)
+print(obj)
+print(theta)
+print(cov)
```
[the remark this answers is not in the document]
Comment (adowling2) [reply pyomo/contrib/parmest/examples/rooney_biegler/parmest_example_cov.py]: Added.

File: pyomo/contrib/interior_point/inverse_reduced_hessian.py [hunk 4]
```diff
@@ -8,7 +8,7 @@
 
 
 # Todo: This function currently used IPOPT for the initial solve - should accept solver
-def inv_reduced_hessian_barrier(model, independent_variables, bound_tolerance=1e-6, tee=False):
+def inv_reduced_hessian_barrier(model, independent_variables, bound_tolerance=1e-6, solver_options=None, tee=False):
```
[the remark this answers is not in the document]
Comment (adowling2) [reply pyomo/contrib/interior_point/inverse_reduced_hessian.py:11]: I agree! Happy to consult with @carldlaird @michaelbynum.

**Comment by jsiirola** [comment]
Note that we are seeing two failing tests on Travis.  These are due to failing documentation builds on platforms where numpy/scipy aren't available.  I believe this is due to the added import of `inv_reduced_hessian_barrier` in `parmest.py`.  This triggers the numpy/scipy checks in the interior_point's `__init__.py`, which raise a `unittest.SkipTest` exception.

@michaelbynum, @Robbybp: I believe that this is actually exposing a problem with interior_point.  I believe that the correct behavior is for the interior point code to raise the `SkipTest` exception in the *test* module and not in the main code.  There should also be an `ImportError` raised (in a class `__init__()` (not at the module scope).

**Comment by michaelbynum** [comment]
I think this means we need to remove everything from the __init__.py in `pyomo.contrib.interior_point` and update all of the interior point tests.

**Comment by Robbybp** [comment]
> I think this means we need to remove everything from the `__init__.py` in `pyomo.contrib.interior_point` and update all of the interior point tests.

We just need to move the checks for numpy/scipy from the `__init__.py` file and make sure the code raises an `ImportError` somewhere if it's used without them, right? The tests all seem to have their own `SkipTest` if numpy and scipy are not available.

**Comment by jsiirola** [comment]
@Robbybp: the problem is that `interior_point/__init__.py` imports classes from both `interior_point` modules and `pynumero` (e.g., `BlockVector`) that unconditionally require numpy/scipy.  This generates an `ImportError` if they are not present.  When `nose` is gathering tests, it attempts to import moduloes from `interior_point/tests/`, which forces the Python importer to import `interior_point/__init__.py`, generating the ImportError.  This error is recorded by `nose` as a test failure.

The general issue is that we probably should not have any `__init__.py` between the `pyomo` namespace and a test module that causes an optional dependency to be forcibly imported.

**Comment by adowling2** [comment]
This is from the failing pypy2 test:
```
======================================================================
ERROR: Failure: ImportError (Error importing asl.Make sure libpynumero_ASL is installed and added to path.)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/opt/hostedtoolcache/PyPy/2.7.13/x64/site-packages/nose/loader.py", line 418, in loadTestsFromName
    addr.filename, addr.module)
  File "/opt/hostedtoolcache/PyPy/2.7.13/x64/site-packages/nose/importer.py", line 47, in importFromPath
    return self.importFromDir(dir_path, fqname)
  File "/opt/hostedtoolcache/PyPy/2.7.13/x64/site-packages/nose/importer.py", line 94, in importFromDir
    mod = load_module(part_fqname, fh, filename, desc)
  File "/home/runner/work/pyomo/pyomo/pyomo/contrib/parmest/tests/test_parmest.py", line 36, in <module>
    import pyomo.contrib.parmest.parmest as parmest
  File "/home/runner/work/pyomo/pyomo/pyomo/contrib/parmest/parmest.py", line 35, in <module>
    from pyomo.contrib.interior_point.inverse_reduced_hessian import inv_reduced_hessian_barrier
  File "/home/runner/work/pyomo/pyomo/pyomo/contrib/interior_point/inverse_reduced_hessian.py", line 4, in <module>
```

**Comment by adowling2** [comment]
@jsiirola @carldlaird I made the conditional import changes we discussed. Fingers crossed this fixes the automated tests.

**Comment by adowling2** [comment]
Here are the details on the failing CI /win/3.7 pull_request test:

```Process code coverage report```

```
    -> Sleeping for 30s and trying again...
    -> Pinging Codecov
https://codecov.io/upload/v4?package=bash-20200602-f809a24&token=secret&branch=parmest-covariance&commit=0fa8a9fd2abab4301294c51b2db128466b3d8bfd&build=138902834&build_url=http%3A%2F%2Fgithub.com%2FPyomo%2Fpyomo%2Factions%2Fruns%2F138902834&name=win%2F3.7&tag=&slug=Pyomo%2Fpyomo&service=github-actions&flags=&pr=1475&job=
    -> Uploading
    X> Failed to upload
    -> Sleeping for 30s and trying again...
    -> Pinging Codecov
https://codecov.io/upload/v4?package=bash-20200602-f809a24&token=secret&branch=parmest-covariance&commit=0fa8a9fd2abab4301294c51b2db128466b3d8bfd&build=138902834&build_url=http%3A%2F%2Fgithub.com%2FPyomo%2Fpyomo%2Factions%2Fruns%2F138902834&name=win%2F3.7&tag=&slug=Pyomo%2Fpyomo&service=github-actions&flags=&pr=1475&job=
    -> Uploading
    X> Failed to upload
    -> Sleeping for 30s and trying again...
    -> Uploading to Codecov
    -> Sleeping for 30s and trying again...
    -> Sleeping for 30s and trying again...
    -> Sleeping for 30s and trying again...
    -> Sleeping for 30s and trying again...
    X> Failed to upload coverage reports
##[error]Process completed with exit code 1.
```

**Comment by blnicho** [comment]
The Windows GitHub Actions job that failed was due to intermittent issue with upload code coverage to codecov. We can go ahead and merge this.

**Pull request merged**